### ISSUE TITLE
Implement support for censored data

### ISSUE THREAD

**Issue by StaffanBetner** [glmmTMB/glmmTMB#690]
Support for censored response data would be useful. The likelihood for the censored data points gets replaced by Pr(a<y≤b) in the general case (interval censoring), which can be written as F(b;θ)-F(a;θ). Whether the upper bound is open or closed does not matter for continuous distributions, but matters for discrete families.

**Comment by bbolker** [comment]
I agree this would be useful. The guts of the computation aren't too hard; [TMB has most of the necessary cumulative distribution functions available already](https://kaskr.github.io/adcomp/group__R__style__distribution.html) although some (`pnbinom`) would have to be written/imported from Rmath.

The hard part is deciding on how the information should be passed to the TMB code.  I can imagine a `censor` argument that names a column that indicates censoring of a given observation (e.g. 0=no censoring, -1=left censoring, 1=right censoring?), and/or a `+censor(...)` element that could be added to the fixed-effect formula (analogous with the existing `weights` argument, although I will say that weights (and offsets) are (a) special-cased in `model.frame()` and (b) kind of a pain to deal with in terms of evaluation. What about interval censoring? In principle one could pass the response variable as a two-column matrix, but this does have the potential to open up all kinds of annoying edge cases ...

If you had a *particular* use case that would be immediately helpful we could start a branch to handle it and begin to explore the issues ...

**Comment by StaffanBetner** [comment]
Maybe the brms syntax is worth copying, although I think it is a bit inflexible with regard to interval censoring. E.g. if the regression distribution is strictly positive, and the interval is 0 to 5, then brms does not handle it as a left-censored 5, as it just fails. But that is just an implementation detail.

Here is from the help page for `brmsformula`:
> With the exception of categorical, ordinal, and mixture families, left, right, and interval censoring can be modeled through y | cens(censored) ~ predictors. The censoring variable (named censored in this example) should contain the values 'left', 'none', 'right', and 'interval' (or equivalently -1, 0, 1, and 2) to indicate that the corresponding observation is left censored, not censored, right censored, or interval censored. For interval censored data, a second variable (let's call it y2) has to be passed to cens. In this case, the formula has the structure y | cens(censored, y2) ~ predictors. While the lower bounds are given in y, the upper bounds are given in y2 for interval censored data. Intervals are assumed to be open on the left and closed on the right: (y, y2].

My use case for this that motivated me to write this issue was to model a visual analogue scale from 0 to 100 (or 1) in steps of 5 (or 0.05) with a beta regression to compare it to a gaussian linear geeglm model with robust standard errors since I was a bit unsure how good the standard errors was in this case due to skewness. The idea was to handle each observation as censored from y-2.5 to y+2.5 (except in the corner cases) to approximate a censored likelihood. I worked around this in glmmTMB by approximating censoring by randomly sample within the interval and then give each sample a weight by 1/n, where n is the number of random samples from the intervals. Seems to work well! (and the differences and their standard errors between the treatment groups calculated with emmeans was pretty much identical to the geeglm setting)

**Comment by bwiernik** [comment]
I would really love if the same syntax used by brms would be adopted for censoring by glmmTMB.

**Comment by jchar32** [comment]
Hoping to check in on this as I have run into a similar need. I am looking to account for right-censored data in a mixed model with a negative binomial distribution. Thanks so much!

**Comment by bbolker** [comment]
How adventurous are you? I might be able to take a preliminary shot at hacking this, if you're willing and able to do some hacking/development yourself. If you need an end-user/turn-key solution, that might take a while longer ... 

Is the censoring point variable by observation or is there a single max/censored value? (The latter would be easier.)

**Comment by jchar32** [comment]
> How adventurous are you? I might be able to take a preliminary shot at hacking this, if you're willing and able to do some hacking/development yourself. If you need an end-user/turn-key solution, that might take a while longer ...
> 
> Is the censoring point variable by observation or is there a single max/censored value? (The latter would be easier.)

@bbolker  Thanks for reaching out so fast! Adventurous, certainly - though I doubt I am qualified to hack/develop (I am open to try, but I haven't worked under the hood of R before). Happy to move to email to discuss further
 
The model I need to fit is fairly simple, two fixed effects, one random intercept on the participant, using a neg binom distribution. 
y ~ variable + group + (1 | participantID)

The dataset is 30 participants with 7 variables extracted for each participant and a binary grouping variable. For some of the 7 variables, participants did not reach the criteria and thus have NA in place of their data point. 

_The censor limit for each participant is different unfortunately_. One strategy to unbias the median estimates from this data was to replace censored data with the known upper limit +1 for that participant. This "uncensored" dataset could perhaps be modelled in the same manner without the need to incorporate censoring into the model structure itself? I imagine this isn't a perfect solution.

**Comment by DrJerryTAO** [comment]
I think the development of interval-censored data models can be combined with "Implement ordinal regression structures" #514 because interval censored responses are a special case of ordinal response where thresholds (can vary by observation) are all known.

**Comment by bbolker** [comment]
From Bryan Maloney:

> I will soon be handed data that has the following features:

> Right-censored.

> Counts (0 to 40+), censored at 40. All zeroes are true zeroes.

> Has offsets of 168 or 175 (There was a "glitch" in the data
collection, where the observation robot recorded for either 168 or 175
hours at a time, no, we can't get the hourly data.)

> Do not know beforehand if data will be overdispersed, underdispersed,
or zero-inflated, so would need to be able to handle all three possibilities.

> The desired test is a test of equivalence.

> The treatments are 4 different specific days of observation.

> Right now we have a gold standard length of observation (in weeks). We
want to test equivalence for using a shorter observation period.

> Now, I may have kludgy ways to do handle all of this, but it would
also be nice if a package that I am very comfortable with, like
glmmTMB were able to answer these questions.

**Comment by DrJerryTAO** [comment]
> From Bryan Maloney:

Such discrete counts with censoring is best handled by cumulative link models (ordered logit model) discussed in #514.

**Comment by BryanJMaloney** [comment]
How would a clmm handle that better than a right-censored poisson or compois model via glMMTMB?
